// ton-core/dist/index.js
"use strict";
/**
 * Copyright (c) Whales Corp.
 * All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeSignVerify = exports.safeSign = exports.getMethodId = exports.base32Encode = exports.base32Decode = exports.crc32c = exports.crc16 = exports.fromNano = exports.toNano = exports.ComputeError = exports.openContract = exports.TupleBuilder = exports.TupleReader = exports.serializeTuple = exports.parseTuple = exports.exoticPruned = exports.exoticMerkleUpdate = exports.exoticMerkleProof = exports.Dictionary = exports.Cell = exports.CellType = exports.Slice = exports.beginCell = exports.Builder = exports.BitBuilder = exports.BitReader = exports.BitString = exports.contractAddress = exports.ADNLAddress = exports.ExternalAddress = exports.address = exports.Address = void 0;
// Address
var Address_1 = require("./address/Address");
Object.defineProperty(exports, "Address", { enumerable: true, get: function () { return Address_1.Address; } });
Object.defineProperty(exports, "address", { enumerable: true, get: function () { return Address_1.address; } });
var ExternalAddress_1 = require("./address/ExternalAddress");
Object.defineProperty(exports, "ExternalAddress", { enumerable: true, get: function () { return ExternalAddress_1.ExternalAddress; } });
var ADNLAddress_1 = require("./address/ADNLAddress");
Object.defineProperty(exports, "ADNLAddress", { enumerable: true, get: function () { return ADNLAddress_1.ADNLAddress; } });
var contractAddress_1 = require("./address/contractAddress");
Object.defineProperty(exports, "contractAddress", { enumerable: true, get: function () { return contractAddress_1.contractAddress; } });
// BitString
var BitString_1 = require("./boc/BitString");
Object.defineProperty(exports, "BitString", { enumerable: true, get: function () { return BitString_1.BitString; } });
var BitReader_1 = require("./boc/BitReader");
Object.defineProperty(exports, "BitReader", { enumerable: true, get: function () { return BitReader_1.BitReader; } });
var BitBuilder_1 = require("./boc/BitBuilder");
Object.defineProperty(exports, "BitBuilder", { enumerable: true, get: function () { return BitBuilder_1.BitBuilder; } });
// Cell
var Builder_1 = require("./boc/Builder");
Object.defineProperty(exports, "Builder", { enumerable: true, get: function () { return Builder_1.Builder; } });
Object.defineProperty(exports, "beginCell", { enumerable: true, get: function () { return Builder_1.beginCell; } });
var Slice_1 = require("./boc/Slice");
Object.defineProperty(exports, "Slice", { enumerable: true, get: function () { return Slice_1.Slice; } });
var CellType_1 = require("./boc/CellType");
Object.defineProperty(exports, "CellType", { enumerable: true, get: function () { return CellType_1.CellType; } });
var Cell_1 = require("./boc/Cell");
Object.defineProperty(exports, "Cell", { enumerable: true, get: function () { return Cell_1.Cell; } });
// Dict
var Dictionary_1 = require("./dict/Dictionary");
Object.defineProperty(exports, "Dictionary", { enumerable: true, get: function () { return Dictionary_1.Dictionary; } });
// Exotics
var exoticMerkleProof_1 = require("./boc/cell/exoticMerkleProof");
Object.defineProperty(exports, "exoticMerkleProof", { enumerable: true, get: function () { return exoticMerkleProof_1.exoticMerkleProof; } });
var exoticMerkleUpdate_1 = require("./boc/cell/exoticMerkleUpdate");
Object.defineProperty(exports, "exoticMerkleUpdate", { enumerable: true, get: function () { return exoticMerkleUpdate_1.exoticMerkleUpdate; } });
var exoticPruned_1 = require("./boc/cell/exoticPruned");
Object.defineProperty(exports, "exoticPruned", { enumerable: true, get: function () { return exoticPruned_1.exoticPruned; } });
var tuple_1 = require("./tuple/tuple");
Object.defineProperty(exports, "parseTuple", { enumerable: true, get: function () { return tuple_1.parseTuple; } });
Object.defineProperty(exports, "serializeTuple", { enumerable: true, get: function () { return tuple_1.serializeTuple; } });
var reader_1 = require("./tuple/reader");
Object.defineProperty(exports, "TupleReader", { enumerable: true, get: function () { return reader_1.TupleReader; } });
var builder_1 = require("./tuple/builder");
Object.defineProperty(exports, "TupleBuilder", { enumerable: true, get: function () { return builder_1.TupleBuilder; } });
// Types
__exportStar(require("./types/_export"), exports);
var openContract_1 = require("./contract/openContract");
Object.defineProperty(exports, "openContract", { enumerable: true, get: function () { return openContract_1.openContract; } });
var ComputeError_1 = require("./contract/ComputeError");
Object.defineProperty(exports, "ComputeError", { enumerable: true, get: function () { return ComputeError_1.ComputeError; } });
// Utility
var convert_1 = require("./utils/convert");
Object.defineProperty(exports, "toNano", { enumerable: true, get: function () { return convert_1.toNano; } });
Object.defineProperty(exports, "fromNano", { enumerable: true, get: function () { return convert_1.fromNano; } });
var crc16_1 = require("./utils/crc16");
Object.defineProperty(exports, "crc16", { enumerable: true, get: function () { return crc16_1.crc16; } });
var crc32c_1 = require("./utils/crc32c");
Object.defineProperty(exports, "crc32c", { enumerable: true, get: function () { return crc32c_1.crc32c; } });
var base32_1 = require("./utils/base32");
Object.defineProperty(exports, "base32Decode", { enumerable: true, get: function () { return base32_1.base32Decode; } });
Object.defineProperty(exports, "base32Encode", { enumerable: true, get: function () { return base32_1.base32Encode; } });
var getMethodId_1 = require("./utils/getMethodId");
Object.defineProperty(exports, "getMethodId", { enumerable: true, get: function () { return getMethodId_1.getMethodId; } });
// Crypto
var safeSign_1 = require("./crypto/safeSign");
Object.defineProperty(exports, "safeSign", { enumerable: true, get: function () { return safeSign_1.safeSign; } });
Object.defineProperty(exports, "safeSignVerify", { enumerable: true, get: function () { return safeSign_1.safeSignVerify; } });
