// @ton-community/sandbox/dist/event/Event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractEvents = void 0;
const extractors = [
    extractAccountCreated,
    extractMessageSent,
    extractAccountDestroyed,
];
function extractEvents(tx) {
    return extractors.map(f => f(tx)).flat();
}
exports.extractEvents = extractEvents;
function doesAccountExist(state) {
    return !(state === 'uninitialized' || state === 'non-existing');
}
function extractAccountCreated(tx) {
    if (!doesAccountExist(tx.oldStatus) && doesAccountExist(tx.endStatus))
        return [{
                type: 'account_created',
                account: tx.inMessage.info.dest,
            }];
    return [];
}
function extractAccountDestroyed(tx) {
    if (doesAccountExist(tx.oldStatus) && !doesAccountExist(tx.endStatus))
        return [{
                type: 'account_destroyed',
                account: tx.inMessage.info.dest,
            }];
    return [];
}
function extractMessageSent(tx) {
    return tx.outMessages.values().flatMap(m => {
        if (m.info.type !== 'internal') {
            return [];
        }
        return [{
                type: 'message_sent',
                from: m.info.src,
                to: m.info.dest,
                value: m.info.value.coins,
                body: m.body,
                bounced: m.info.bounced,
            }];
    });
}
